files/root/scripts/puppet-profile_additional_packages-packages_without_params.sh

```diff
@@ -0,0 +1,23 @@
+#/bin/bash
+
+# puppet-profile_additional_packages-packages_without_params.sh: Managed by Puppet
+
+# usage: "$0" pkg1 pkg2 pkg3 ...
+
+# installs only the packages that are not already installed and logs those to syslog
+
+OUTPUT_FILE="/root/puppet-profile_additional_packages-packages_without_params.out"
+
+if [ 0 -eq "$#" ]; then
+  /usr/bin/logger -t puppet-profile_additional_packages -p local4.info "script called with no args [no pkg list]"
+  exit 1
+fi
+
+/usr/bin/rm -f ${OUTPUT_FILE}
+/usr/bin/rpm -q $* | /usr/bin/grep 'is not installed' | /usr/bin/awk '{print $2}' > ${OUTPUT_FILE}
+# if the file exists and is NOT zero size
+if [ -f ${OUTPUT_FILE} ] && [ -s ${OUTPUT_FILE} ]; then
+  install_packages=`/usr/bin/cat ${OUTPUT_FILE} | tr '\n' ' '`
+  /usr/bin/logger -t puppet-profile_additional_packages -p local4.info "installing ${install_packages}"
+  /usr/bin/dnf -y install ${install_packages}
```

Comment:
We want to make sure that this script returns the error code from this command.

Reply:
Yep, if it gets to that statement it returns the exit code of dnf. And it does report some useful output:
`Error: '/root/scripts/puppet-profile_additional_packages-packages_without_params.sh bzip2 git iotop lsof net-tools nmap rsync strace tcpdump tcsh traceroute tree vim-enhanced wget zsh htop mailx hklkldfj apptainer asciidoc autoconf automake binutils binutils-devel bison blas-devel boost-devel byacc cmake createrepo_c ctags diffstat dos2unix flex gcc gdb geos geos-devel glibc-devel gsl-devel hwloc ImageMagick-c++-devel intltool java-1.8.0-openjdk java-1.8.0-openjdk-devel java-11-openjdk-devel jna lapack-devel libevent-devel libffi-devel libgsf-devel libjwt librsvg2-devel libsodium-devel libtiff-devel libtool libusbx ltrace lz4 make nano numactl numactl-devel openjpeg2-devel openssl-devel patchutils pcre2-devel perl-Fedora-VSP perl-generators pesign pkgconf pkgconf-m4 pkgconf-pkg-config postgresql-odbc proj proj-devel protobuf-compiler protobuf-devel python3-tabulate redhat-rpm-config rpm-build rpm-sign rrdtool source-highlight sqlite sqlite-devel udunits2-devel unixODBC-devel valgrind valgrind-devel xorg-x11-apps xorg-x11-xauth zenity' returned 1 instead of one of [0]
Error: /Stage[main]/Profile_additional_packages/Exec[install_packages_without_params]/returns: change from 'notrun' to ['0'] failed: '/root/scripts/puppet-profile_additional_packages-packages_without_params.sh bzip2 git iotop lsof net-tools nmap rsync strace tcpdump tcsh traceroute tree vim-enhanced wget zsh htop mailx hklkldfj apptainer asciidoc autoconf automake binutils binutils-devel bison blas-devel boost-devel byacc cmake createrepo_c ctags diffstat dos2unix flex gcc gdb geos geos-devel glibc-devel gsl-devel hwloc ImageMagick-c++-devel intltool java-1.8.0-openjdk java-1.8.0-openjdk-devel java-11-openjdk-devel jna lapack-devel libevent-devel libffi-devel libgsf-devel libjwt librsvg2-devel libsodium-devel libtiff-devel libtool libusbx ltrace lz4 make nano numactl numactl-devel openjpeg2-devel openssl-devel patchutils pcre2-devel perl-Fedora-VSP perl-generators pesign pkgconf pkgconf-m4 pkgconf-pkg-config postgresql-odbc proj proj-devel protobuf-compiler protobuf-devel python3-tabulate redhat-rpm-config rpm-build rpm-sign rrdtool source-highlight sqlite sqlite-devel udunits2-devel unixODBC-devel valgrind valgrind-devel xorg-x11-apps xorg-x11-xauth zenity' returned 1 instead of one of [0] (corrective)
Error: Execution of '/usr/bin/dnf -d 0 -e 1 -y install hklkldfj' returned 1: Error: Unable to find a match: hklkldfj
Error: /Stage[main]/Profile_additional_packages/Package[hklkldfj]/ensure: change from 'purged' to 'present' failed: Execution of '/usr/bin/dnf -d 0 -e 1 -y install hklkldfj' returned 1: Error: Unable to find a match: hklkldfj
`

You can see above that it's now trying via exec first, where it shows:

1. the full script command w/ the entire list of param-less pkgs passed as args to the script,
2. and then shows the actual dnf command that ran and errored,
3. and then finally tries a package resource (from ensure_packages) and fails that too.

I can't quite tell if #2 is associated w/ #3 or w/ #1. (But again, syslog will show the exact packages that the exec tried to install too.)

(I did update the script to use dnf, but I guess that ultimately points to yum on RHEL 8 and hence logged that way.)

Note that I updated the yum. command in the script by adding '-d 0 -e 1', which Puppet itself uses, to suppress some of the more noisy output of yum.

Reply:
Cool